pubtools/pulplib/_impl/hooks.py

```diff
@@ -6,14 +6,26 @@
 # pylint: disable=unused-argument
 
 
+@hookspec
+def pulp_repository_pre_publish(to_publish):
```

Comment:
I'm not comfortable with some things in the design here:

- the mixing of immutable high-level objects (Distributor) and mutable low-level raw pulp data (override_config) in the same function.
- how different the signature is to the similarly named pulp_repository_published
- the fact that the caller needs to know about the format of override_config, and will then end up relying on pulp2-specific implementation details which we don't document in pulplib.

Could we please make it higher-level instead?  I think the better way might be like this:

- Explicitly add rsync_extra_args to PublishOptions class, with docs along the lines of "If present, provide these additional arguments to any rsync commands run during publish.  Ignored when rsync is not used."
- Implement support internally in pulplib so that if options.rsync_extra_args is set, we adjust override_config to include that for cdn_distributor.
- Make the hook signature like this: `def pulp_repository_pre_publish(repository, options):`
- Define the hook as returning a PublishOptions object.  In the doc string, state that "If a hookimpl returns a non-None value, that value will be used to replace the options for this publish.  This can be used to adjust publish options from within a hook."
- Implement the behavior matching that doc string (i.e. check the return value of the hook)

This keeps us using only the documented high-level API and also gives us a better chance of somehow keeping it working if we move to pulp3.